crates/node/src/roles/worker.rs

```diff
@@ -118,6 +128,89 @@ impl WorkerRole {
+                let mut first_attempt = true;
+
+                let metadata_client = metadata_writer.global_metadata();
+
+                if let Err(err) = retry_on_retryable_error(retry_policy, || {
+                    metadata_client.read_modify_write(
+                        move |nodes_config: Option<Arc<NodesConfiguration>>| {
+                            let nodes_config = nodes_config.expect(
+                                "nodes config must be present if the node starts the worker role",
+                            );
+
+                            let node_config =
+                                nodes_config.find_node_by_id(my_node_id).map_err(|err| {
+                                    ProvisionWorkerRoleError::NewerGenerationOrRemoved(err)
+                                })?;
+
+                            if node_config.worker_config.worker_state != WorkerState::Provisioning {
+                                return if first_attempt {
+                                    Err(ProvisionWorkerRoleError::NotProvisioning(
+                                        node_config.worker_config.worker_state,
+                                    ))
+                                } else {
+                                    Err(ProvisionWorkerRoleError::PreviousAttemptSucceeded)
+                                };
+                            }
+
+                            let mut my_node_config = node_config.clone();
+                            my_node_config.worker_config.worker_state = WorkerState::Active;
+
+                            let mut new_nodes_config = nodes_config.as_ref().clone();
+                            new_nodes_config.upsert_node(my_node_config);
+                            new_nodes_config.increment_version();
+
+                            first_attempt = false;
+
+                            Ok(new_nodes_config)
+                        },
+                    )
+                })
+                .await
+                .map_err(|err| err.map(|err| err.transpose()))
+                {
+                    match err {
+                        RetryError::RetriesExhausted(
+                            ProvisionWorkerRoleError::PreviousAttemptSucceeded,
+                        )
+                        | RetryError::NotRetryable(
+                            ProvisionWorkerRoleError::PreviousAttemptSucceeded,
+                        ) => {}
+                        err => {
+                            bail!("failed to update worker state: {}", err);
+                        }
+                    }
+                }
+            }
+            WorkerState::Active | WorkerState::Draining => {
+                // nothing to do if we are already active or draining
+            }
+            WorkerState::Disabled => {
+                bail!(
+                    "Worker role is disabled and should no longer be started on this node {my_node_id}"
+                );
+            }
+        }
```

Comment:
I'm not entirely sure why we'd want to reject starting up the node if it worker has been disabled. What are we risking?
- Is it sufficient to let the work role start and be quiescent, PPM running, but no partitions are placed on this node anyway.
- Is there a risk of allowing the node with the same plain node ID from being put back to Provisioning or Active after disable?

Reply:
I think you are right that this is too conservative. Removing the bail to not crash the node in this case.